### DISCUSSION TITLE
Load Testing

### DISCUSSION THREAD

**Discussion by nicnocquee** [hyperjumptech/monika#284] [Ideas]
We need to know how well Monika can perform under heavy loads, i.e., a lot of probes and requests. Few questions that need answers:

 1. As the number of probes grow in thousands, how much memory, cpu, and disk usage used?
 2. As the number of probes with chained requests grow in thousands, how much memory, cpu, and disk usage used?
 3. As the number of probes grow in thousands, are there dropped requests?

cc @samsunoor

**Comment by syamsudotdev** [comment]
Quick test on ~512MB machine, running Monika configured with ~1200 HTTP GET requests. Memory usage will be full almost instantly. Thus resulting Monika got killed silently by the OS, possibly OOM.

This is the [monika.json.txt](https://github.com/hyperjumptech/monika/files/6840633/monika.json.txt) configuration file that I use.

Edit
Same thing will happen with ~500 requests. Will try to find the threshold later from low number of requests.

**Reply by syamsudotdev** [reply]
After further checking, it never went OOM.
Starting from 300 requests, Monika will use ~175 MB of memory at peak with not-fully-utilized CPU usage. On that peak for a while, the process suddenly get killed.
The more number of requests, the faster it will get killed. I tested with the 1K requests, it got killed around the first minute.

**Comment by syamsudotdev** [comment]
Edit:
Test machine (X1 Carbon 5th Gen) specs:
- `i7-7500U`
- 8 GB RAM
- Storage SSD NVMe
- Kernel: `5.10.49-1-MANJARO #1 SMP PREEMPT Sun Jul 11 12:59:43 UTC 2021 x86_64 GNU/Linux`
---
Tested with [1K requests](https://github.com/hyperjumptech/monika/files/6866889/monika.json.txt), [5K requests](https://github.com/hyperjumptech/monika/files/6866892/monika.json.txt), and [10K requests](https://github.com/hyperjumptech/monika/files/6866900/monika.json.txt) configuration.

#### As the number of probes grow in thousands, how much memory, cpu, and disk usage used?
- With 1K requests and lower will need memory under 400MB running for ~5 hours.
- Starting with 5K requests and up, running for ~5 hours, Monika need around 400MB of memory on average. It could spike up to 2.5GB sometimes.
- Disk usage for ~3 hours monitoring will need ~150MB of space for the SQLite file.
- On `i7-7500U`, a Monika instance generally need lower than 5% CPU, sometimes it spiked to ~10%.
- Splitting 10K requests to 5K requests config does not make the numbers above lower per instance. Maybe they need to run for a few days to get different results.

#### As the number of probes with chained requests grow in thousands, how much memory, cpu, and disk usage used?
- I think the previous point still stand because the storage works in same manner but the CPU for string interpolation process.

#### As the number of probes grow in thousands, are there dropped requests?
- There were no errors on the output log nor dropped requests up to 10K requests configuration.

**Reply by sapiderman** [reply]
Syam, what was your teset procedure? 
What was your monika.json configuration? 
Are you saving the body or not?

**Reply by syamsudotdev** [reply]
- Did tests by running Monika with different config for 3-6 hours each. Some tests ran by one Monika instance. Some tests ran 3 Monika instance with `2 instances x 5K requests + 1 instance x 10K requests`
- Configurations are included above.
- Did not save the response body.
